study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/translation/ActivityDefTranslationsProcessor.java

```diff
@@ -0,0 +1,188 @@
+import org.broadinstitute.ddp.model.activity.definition.MailingAddressComponentDef;
+import org.broadinstitute.ddp.model.activity.definition.NestedActivityBlockDef;
+import org.broadinstitute.ddp.model.activity.definition.PhysicianInstitutionComponentDef;
+import org.broadinstitute.ddp.model.activity.definition.QuestionBlockDef;
+import org.broadinstitute.ddp.model.activity.definition.question.QuestionDef;
+import org.broadinstitute.ddp.model.activity.definition.validation.RuleDef;
+import org.broadinstitute.ddp.studybuilder.StudyBuilderContext;
+import org.broadinstitute.ddp.studybuilder.StudyBuilderException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Add the translations (for all study languages) to the activities definitions.
+ * Translations can be defined as an array in variables which defined in Templates.
+ * Or Translations can be defined as an array in activity definition (like `translatedNames`, `translatedSecondNames`).
+ *
+ * <p><b>Algorithm:</b>
+ * <ul>
+ *     <li>detect active languages;</li>
+ *     <li>process all activity-level templates;</li>
+ *     <li>go through all sections and process section level templates;</li>
+ *     <li>go through all blocks of a section and process all block level templates (depending on a block type);</li>
+ *     <li>for each template do the following steps:
+ *       <ol>
+ *           <li>if a translation without template (like activityDef.translatedTitles, activityDef.translatedSubtitles)
+ *               then process only translations arrays which
+ *               are not empty and contain at least one translation (for example with default language):
+ *               detect from such element a name of a translation and add translations for the rest of languages
+ *               (which are still not defined);</li>
+ *           <li>if a template is processed then check it's variables:
+ *               - if a variable contains non-empty translation array containing at least one translation
+ *                 (for example with default language):
+ *                 detect from such element a name of a translation and add translations for the rest of languages
+ *                 (which are still not defined);
+ *               - if variables array is empty then detect names of variables from a Template text and build
+ *                 translations for each of detected variables (for each of active languages).
+ *           </li>
+ *       </ol>
+ *     </li>
+ * </ul>
+ */
+public class ActivityDefTranslationsProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ActivityDefTranslationsProcessor.class);
+
+    private final Map<String, Properties> allTranslations;
+
+    public ActivityDefTranslationsProcessor(Map<String, Properties> allTranslations) {
+        this.allTranslations = allTranslations;
+    }
+
+    public void run(ActivityDef activityDef) {
+        if (StudyBuilderContext.CONTEXT.isProcessTranslations()) {
```

Comment:
Translations processing will start only if context.processTranslations==true (which is assigned by cmd option `--process-translations`).